### PR TITLE
feat(auth): support RFC 8414 and OIDC Discovery for auth server metadata

### PR DESCRIPTION
**Pull request by gocamille**
#### Summary 

Update authorization server metadata discovery to support both OAuth 2.0 AS Metadata (RFC 8414) and OpenID Connect Discovery 1.0, per [MCP Auth Spec 2025-11-25](https://modelcontextprotocol.io/specification/2025-11-25/basic/authorization#authorization-server-metadata-discovery).

#### Background

The previous implementation only attempted a single discovery URL pattern. The latest spec requires trying multiple URL patterns to ensure compatibility with various OAuth providers:

- RFC 8414 (path-insertion): `/.well-known/oauth-authorization-server/{path}`
- OIDC Discovery (path-insertion): `/.well-known/openid-configuration/{path}`
- OIDC Discovery (legacy path-appending): `/{path}/.well-known/openid-configuration`

#### Changes

- New `build_discovery_urls` function which builds discovery URLs in priority order per MCP spec
- New `discover_jwks` function 
- URL normalization to strip query strings/fragments per [RFC 8414 Section 3.1](https://datatracker.ietf.org/doc/html/rfc8414#section-3.1), which requires us to collapse double slashes

### CODE REVIEW

File: crates/apollo-mcp-server/src/auth/networked_token_validator.rs [hunk 1]
```diff
@@ -29,15 +31,101 @@ impl<'a> NetworkedTokenValidator<'a> {
...
+        .collect()
+}
+
+/// Attempts discovery from multiple URLs sequentially, returning first success
+async fn discover_jwks(client: &reqwest::Client, issuer: &Url) -> Option<Jwks> {
+    let urls = build_discovery_urls(issuer);
+
+    for url in &urls {
+        let result = tokio::time::timeout(
+            Duration::from_secs(5),
```
Comment (andrewmcgivery) [inline crates/apollo-mcp-server/src/auth/networked_token_validator.rs]: Bit of a magic number here.... should this be configurable? I know in Router land stuff life this often gets requested to be configurable.

Comment (gocamille) [reply crates/apollo-mcp-server/src/auth/networked_token_validator.rs]: This is a good point! I'm always weary of these and have seen a few of these magic numbers in the repo, I would tend to agree we should expose this to be configurable 👍

File: crates/apollo-mcp-server/src/auth/networked_token_validator.rs [hunk 2]
```diff
@@ -29,15 +31,101 @@ impl<'a> NetworkedTokenValidator<'a> {
...
-    discovery_url
+/// Constructs discovery URLs. Returns URLs in priority order:
+/// 1. RFC 8414 (path-insertion): `/.well-known/oauth-authorization-server/{path}`
+/// 2. OIDC Discovery (path-insertion): `/.well-known/openid-configuration/{path}`
+/// 3. OIDC Discovery (legacy path-appending): `/{path}/.well-known/openid-configuration`
+///
+/// # URL Normalization
+/// Query strings and fragments are stripped per RFC 8414 Section 3.1.
+/// The spec does not define behavior for these, and most implementations strip them.
+fn build_discovery_urls(issuer: &Url) -> Vec<Url> {
```
Comment (DaleSeo) [inline crates/apollo-mcp-server/src/auth/networked_token_validator.rs]: I like the new function and parameter name. 👍

File: crates/apollo-mcp-server/src/auth/networked_token_validator.rs [hunk 3]
```diff
@@ -29,15 +31,101 @@ impl<'a> NetworkedTokenValidator<'a> {
...
+            origin, path_suffix
+        ),
+        // Priority 2: OIDC Discovery path-insertion
+        format!("{}/.well-known/openid-configuration{}", origin, path_suffix),
+    ];
+
+    // Priority 3: OIDC Discovery legacy path-appending (only if there IS a path)
+    if !path.is_empty() {
+        urls.push(format!(
+            "{}/{}/.well-known/openid-configuration",
```
Comment (DaleSeo) [inline crates/apollo-mcp-server/src/auth/networked_token_validator.rs:100]: I noticed the fallback behavior changed slightly. Previously, the path-appending style used oauth-authorization-server:

```rs
{path}/.well-known/oauth-authorization-server
```

But now the path-appending fallback (priority 3) uses openid-configuration:

```rs
{path}/.well-known/openid-configuration
```

If any customers have been relying on the previous discovery behavior, this could be a breaking change for them.

Should we consider adding `{path}/.well-known/oauth-authorization-server` as a 4th fallback for backwards compatibility? Curious if you've tested against the providers we were previously supporting.

File: crates/apollo-mcp-server/src/auth/networked_token_validator.rs [hunk 4]
```diff
@@ -29,15 +31,101 @@ impl<'a> NetworkedTokenValidator<'a> {
...
+    let path = normalized
+        .path()
+        .trim_matches('/')
+        .split('/')
+        .filter(|s| !s.is_empty())
+        .collect::<Vec<_>>()
+        .join("/");
+
+    let Some(host) = normalized.host_str() else {
+        trace!(issuer = %issuer, "Issuer URL has no host, cannot build discovery URLs");
```
Comment (DaleSeo) [inline crates/apollo-mcp-server/src/auth/networked_token_validator.rs]: I'm a bit concerned that this trace log might go unnoticed in practice. Most users won't run at trace level unless they're actively debugging. If someone accidentally configures a URL without a host, auth will silently fail and the only hint will be buried in trace logs.

A URL without a host should be a clear configuration mistake rather than an expected edge case. Would it make sense to validate this at config load time instead? That would give users a clear startup error rather than a runtime mystery.

File: .gitignore [hunk 5]
```diff
@@ -9,6 +9,7 @@
 # will have compiled files and executables
 debug/
 target/
+target-install/
```
Comment (DaleSeo) [inline .gitignore]: What is this for?

File: .changeset/support_oauth_20_and_openid_connect_metadata_discovery.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+default: minor
+---
+
+# Support OAuth 2.0 and OpenID Connect metadata discovery
```
Comment (DaleSeo) [inline .changeset/support_oauth_20_and_openid_connect_metadata_discovery.md:5]:
```suggestion
### Support OAuth 2.0 and OpenID Connect metadata discovery - @gocamille PR #584

// description
```

Comment (andrewmcgivery) [reply .changeset/support_oauth_20_and_openid_connect_metadata_discovery.md:5]: I wouldn't mess with the heading level. Knope makes it a h1 on purpose and will convert it to the proper heading level in changelogs later as part of it's release process.

File: crates/apollo-mcp-server/src/auth/networked_token_validator.rs (hunk 3 above)
Comment (gocamille) [reply crates/apollo-mcp-server/src/auth/networked_token_validator.rs:100]: @DaleSeo Good catch and I did look into this while building! I noticed the comment in commit [1fcd0fc](https://github.com/apollographql/apollo-mcp-server/commit/1fcd0fc) for Keycloak specifically. The fallback order change is intentional per MCP Auth Spec 2025-11-25.

For Keycloak, which was the vendor referenced in our project, this is not a breaking change. Keycloak serves metadata at `{realm}/.well-known/openid-configuration`, which is exactly what [Priority 3](https://modelcontextprotocol.io/specification/2025-11-25/basic/authorization#authorization-server-metadata-discovery) hits. The previous `oauth-authorization-server` path-appending only worked because some Keycloak versions serve both endpoints—and `openid-configuration` is the canonical one 👍 

One note though: Keycloak users will see 2 failed requests (Priority 1-2 are path-insertion URLs Keycloak doesn't serve) before Priority 3 succeeds. But the `404`s are fast (~50ms each), so total overhead is ~100ms.

I'm happy to add a Priority 4 for `{path}/.well-known/oauth-authorization-server` if we want strict backwards compatibility, though it's not spec-mandated and Keycloak doesn't need it.

File: crates/apollo-mcp-server/src/auth/networked_token_validator.rs (hunk 4 above)
Comment (gocamille) [reply crates/apollo-mcp-server/src/auth/networked_token_validator.rs]: @DaleSeo you're right! This should be a startup error, not a trace log. A host-less URL is a config mistake, not a runtime edge case 👍 

I've added validation in [5d8c2bf](https://github.com/apollographql/apollo-mcp-server/pull/584/commits/5d8c2bfd4e8863e83bcfe87a75db37624af2a649) to reject auth server URLs without hosts. The new `build_discovery_urls()` can then assume valid input.

File: .gitignore (hunk 5 above)
Comment (gocamille) [reply .gitignore]: I was testing a target install and accidentally committed this directory so added it to `.gitignore`! However this is a rare occurrence and mistake, so I am happy to remove this!

File: .changeset/support_oauth_20_and_openid_connect_metadata_discovery.md (hunk 6 above)
Comment (gocamille) [reply .changeset/support_oauth_20_and_openid_connect_metadata_discovery.md:5]: Things I learn!

File: crates/apollo-mcp-server/src/auth/networked_token_validator.rs (hunk 4 above)
Comment (DaleSeo) [reply crates/apollo-mcp-server/src/auth/networked_token_validator.rs]: I get this from the link provided:

2026-01-22 at 14 33 06

Comment (DaleSeo) [reply crates/apollo-mcp-server/src/auth/networked_token_validator.rs]: NVM, I found the validation logic.

File: crates/apollo-mcp-server/src/auth.rs [hunk 7]
```diff
@@ -42,6 +42,12 @@ pub enum TlsConfigError {
     CertificateParse { path: PathBuf },
     #[error("Failed to build HTTP client: {0}")]
     ClientBuild(#[from] reqwest::Error),
+    #[error("Invalid auth server URL at index {index} ({url}): {reason}")]
+    InvalidServerUrl {
+        index: usize,
+        url: String,
+        reason: String,
```
Comment (DaleSeo) [inline crates/apollo-mcp-server/src/auth.rs]: Is the `reason` field intended for future validation cases? Currently it's only ever "URL has no host". If there are no other validations planned, a simpler variant like `ServerUrlMissingHost { index, url }` might be cleaner, but happy to leave as-is if you have other checks in mind.

**Comment by gocamille** [comment]
## Review Summary

This PR has already received a thorough review. I've analyzed the code against Apollo's Rust best practices and found that the previous review comprehensively covered the main issues.

### Confirmed Findings

The previous review correctly identified:
- **Line 56**: The `expect()` usage with SAFETY comment could be more robust (Chapter 4.2)
- **Test assertions**: Could use more descriptive error messages for easier debugging
- **Test naming**: Generally follows conventions well

### Additional Observations

**Performance consideration**: The `discover_jwks` function (lines 102-126) applies a 5-second timeout **per discovery URL**. With 2-3 URLs in the fallback chain, the cumulative worst-case timeout is **10-15 seconds** if all URLs fail sequentially. This is likely acceptable for a one-time initialization operation, but worth considering if this code path is hit per-request (though JWKS caching would typically prevent this).

### Assessment

**Strengths:**
- ✅ Excellent test coverage with comprehensive edge cases (URL normalization, provider patterns, fallback behavior, error handling)
- ✅ Clean implementation of RFC 8414 and OIDC Discovery spec with proper fallback logic
- ✅ Appropriate error logging with trace/warn levels
- ✅ Proper URL normalization (query strings, fragments, path handling)
- ✅ Early validation of server URLs at startup to fail fast

**Test Coverage**: ✅ Excellent - covers edge cases, error paths, and integration scenarios

**Recommendation**: **Approve with suggestions**

The implementation is solid and production-ready. The issues raised in the previous review are minor and don't block merging. The cumulative timeout consideration is worth noting but likely not a practical concern.

---
_Reviewed by Claude Code Sonnet 4.5_

File: crates/apollo-mcp-server/src/auth/networked_token_validator.rs (hunk 3 above)
Comment (DaleSeo) [reply crates/apollo-mcp-server/src/auth/networked_token_validator.rs:100]: Thanks for the thorough investigation, @gocamille! If you're confident, I'm okay skipping Priority 4. Just in case someone hits an edge case, I'd suggest we highlight this behavior change in the changeset file. Something like noting that the discovery order changed and users relying on legacy `{path}/.well-known/oauth-authorization-server` should verify their provider supports `openid-configuration`. 

For context, we tend to keep features deprecated rather than removed since the MCP spec changes quickly, and users often struggle to keep up. For example, we kept SSE transport around for 10 months after it was dropped from the spec. So I'm a bit cautious about removing fallback behaviors, but I'll defer to your judgment here since you've tested it.

File: .changeset/support_oauth_20_and_openid_connect_metadata_discovery.md (hunk 6 above)
Comment (DaleSeo) [reply .changeset/support_oauth_20_and_openid_connect_metadata_discovery.md:5]: Sorry about the incorrect feedback, @gocamille! 😅

File: crates/apollo-mcp-server/src/auth/networked_token_validator.rs [hunk 8]
```diff
@@ -19,25 +22,116 @@ impl<'a> NetworkedTokenValidator<'a> {
...
+    let path = normalized
+        .path()
+        .trim_matches('/')
+        .split('/')
+        .filter(|s| !s.is_empty())
+        .collect::<Vec<_>>()
+        .join("/");
+
+    let Some(host) = normalized.host_str() else {
+        unreachable!("server URL must have a host (validated at startup)")
```
[the remark this answers is not in the document]
Comment (andrewmcgivery) [reply crates/apollo-mcp-server/src/auth/networked_token_validator.rs]: We should probably in general avoid having panic-able code in our codebase. I'd rather double up validation and have this method return a Result<Vec<Url>> than have a `unreachable` or `expect`.

[the remark this answers is not in the document]
Comment (gocamille) [reply crates/apollo-mcp-server/src/auth/networked_token_validator.rs]: Good point @andrewmcgivery (and claude 😂 )! I've updated this on [5d778c7](https://github.com/apollographql/apollo-mcp-server/pull/584/commits/5d778c71e03fbbc0c6aec1b9aa4e511f73cf3bc4)

**Comment by andrewmcgivery** [comment]
@gocamille one small thing I just thought about... probably want to update the docs with the new configuration option!

File: crates/apollo-mcp-server/src/auth.rs (hunk 7 above)
Comment (gocamille) [reply crates/apollo-mcp-server/src/auth.rs]: @DaleSeo you're right that currently there's only one validation case! I kept the reason field for extensibility in case we add other URL validations in the future (e.g., unsupported schemes like `file://` or malformed paths, etc.), but I agree it's speculative.

For now I've gone ahead and simplified it based on your suggestions 👍 !
